File: backend/app/schemas/reporting.py
"""
Reporting schemas for API request/response validation.
"""

from datetime import datetime
from enum import StrEnum

from pydantic import Field

from .base import BaseSchema


class ReportCategorySchema(StrEnum):
    """Report template categories."""

    EXECUTIVE = "executive"
    FINANCIAL = "financial"
    MARKET = "market"
    PORTFOLIO = "portfolio"
    CUSTOM = "custom"


class ReportFormatSchema(StrEnum):
    """Report export formats."""

    PDF = "pdf"
    EXCEL = "excel"
    PPTX = "pptx"


class ReportStatusSchema(StrEnum):
    """Queued report status."""

    PENDING = "pending"
    GENERATING = "generating"
    COMPLETED = "completed"
    FAILED = "failed"


class ScheduleFrequencySchema(StrEnum):
    """Distribution schedule frequency."""

    DAILY = "daily"
    WEEKLY = "weekly"
    MONTHLY = "monthly"
    QUARTERLY = "quarterly"


# ==================== Report Template Schemas ====================


class ReportTemplateBase(BaseSchema):
    """Base schema for report template."""

    name: str = Field(..., min_length=1, max_length=255, description="Template name")
    description: str | None = Field(None, description="Template description")
    category: ReportCategorySchema = Field(
        ReportCategorySchema.CUSTOM,
        description="Template category",
    )
    sections: list[str] = Field(default_factory=list, description="Report sections")
    export_formats: list[ReportFormatSchema] = Field(
        default_factory=lambda: [ReportFormatSchema.PDF],
        description="Available export formats",
    )
    is_default: bool = Field(False, description="Is this a default system template")


class ReportTemplateCreate(ReportTemplateBase):
    """Schema for creating a report template."""

    created_by: str = Field("System", description="User who created the template")
    config: dict | None = Field(None, description="Custom widget/layout configuration")


class ReportTemplateUpdate(BaseSchema):
    """Schema for updating a report template."""

    name: str | None = Field(None, min_length=1, max_length=255)
    description: str | None = None
    category: ReportCategorySchema | None = None
    sections: list[str] | None = None
    export_formats: list[ReportFormatSchema] | None = None
    is_default: bool | None = None
    config: dict | None = None


class ReportTemplateResponse(ReportTemplateBase):
    """Response schema for report template."""

    id: int
    created_by: str
    config: dict | None = None
    created_at: datetime
    updated_at: datetime


class ReportTemplateListResponse(BaseSchema):
    """Response schema for list of report templates."""

    items: list[ReportTemplateResponse]
    total: int
    page: int
    page_size: int


# ==================== Queued Report Schemas ====================


class QueuedReportBase(BaseSchema):
    """Base schema for queued report."""

    name: str = Field(..., min_length=1, max_length=255, description="Report name")
    template_id: int = Field(..., description="Template ID")
    format: ReportFormatSchema = Field(
        ReportFormatSchema.PDF, description="Export format"
    )
    requested_by: str = Field(..., description="User who requested the report")


class QueuedReportCreate(QueuedReportBase):
    """Schema for creating a queued report."""

    pass


class QueuedReportResponse(QueuedReportBase):
    """Response schema for queued report."""

    id: int
    template_name: str | None = Field(None, description="Template name")
    status: ReportStatusSchema
    progress: int = Field(..., ge=0, le=100, description="Generation progress 0-100")
    requested_at: datetime
    completed_at: datetime | None = None
    file_size: str | None = None
    download_url: str | None = None
    error: str | None = None
    created_at: datetime
    updated_at: datetime


class QueuedReportListResponse(BaseSchema):
    """Response schema for list of queued reports."""

    items: list[QueuedReportResponse]
    total: int
    page: int
    page_size: int


# ==================== Distribution Schedule Schemas ====================


class DistributionScheduleBase(BaseSchema):
    """Base schema for distribution schedule."""

    name: str = Field(..., min_length=1, max_length=255, description="Schedule name")
    template_id: int = Field(..., description="Template ID")
    recipients: list[str] = Field(..., min_length=1, description="Email recipients")
    frequency: ScheduleFrequencySchema = Field(
        ..., description="Distribution frequency"
    )
    day_of_week: int | None = Field(
        None, ge=0, le=6, description="Day of week (0=Mon, 6=Sun)"
    )
    day_of_month: int | None = Field(
        None, ge=1, le=31, description="Day of month (1-31)"
    )
    time: str = Field(
        ...,
        pattern=r"^([01]\d|2[0-3]):([0-5]\d)$",
        description="Time in HH:MM format",
    )
    format: ReportFormatSchema = Field(
        ReportFormatSchema.PDF, description="Export format"
    )
    is_active: bool = Field(True, description="Is schedule active")


class DistributionScheduleCreate(DistributionScheduleBase):
    """Schema for creating a distribution schedule."""

    next_scheduled: datetime = Field(..., description="Next scheduled run time")


class DistributionScheduleUpdate(BaseSchema):
    """Schema for updating a distribution schedule."""

    name: str | None = Field(None, min_length=1, max_length=255)
    recipients: list[str] | None = None
    frequency: ScheduleFrequencySchema | None = None
    day_of_week: int | None = None
    day_of_month: int | None = None
    time: str | None = Field(None, pattern=r"^([01]\d|2[0-3]):([0-5]\d)$")
    format: ReportFormatSchema | None = None
    is_active: bool | None = None
    next_scheduled: datetime | None = None


class DistributionScheduleResponse(DistributionScheduleBase):
    """Response schema for distribution schedule."""

    id: int
    template_name: str | None = Field(None, description="Template name")
    last_sent: datetime | None = None
    next_scheduled: datetime
    created_at: datetime
    updated_at: datetime


class DistributionScheduleListResponse(BaseSchema):
    """Response schema for list of distribution schedules."""

    items: list[DistributionScheduleResponse]
    total: int


# ==================== Report Generation Schemas ====================


class GenerateReportRequest(BaseSchema):
    """Request schema for generating a report."""

    template_id: int = Field(..., description="Template ID to use")
    name: str = Field(..., min_length=1, max_length=255, description="Report name")
    format: ReportFormatSchema = Field(
        ReportFormatSchema.PDF, description="Export format"
    )
    parameters: dict | None = Field(None, description="Additional report parameters")


class GenerateReportResponse(BaseSchema):
    """Response schema for report generation request."""

    queued_report_id: int
    status: ReportStatusSchema
    message: str


# ==================== Report Widget Schemas ====================


class ReportWidgetSchema(BaseSchema):
    """Schema for report widget definition."""

    id: str
    type: str = Field(
        ..., description="Widget type: chart, table, metric, text, image, map"
    )
    name: str
    description: str
    category: str
    icon: str
    default_width: int = Field(..., ge=1, le=12, description="Grid width units")
    default_height: int = Field(..., ge=1, description="Grid height units")
    configurable: bool = True


class ReportWidgetListResponse(BaseSchema):
    """Response schema for list of report widgets."""

    widgets: list[ReportWidgetSchema]
    total: int


# ==================== Report Settings Schemas ====================


class ReportSettingsSchema(BaseSchema):
    """Schema for report settings."""

    company_name: str
    company_logo: str | None = None
    primary_color: str
    secondary_color: str
    default_font: str
    default_page_size: str = Field(..., pattern="^(letter|a4|legal)$")
    default_orientation: str = Field(..., pattern="^(portrait|landscape)$")
    include_page_numbers: bool = True
    include_table_of_contents: bool = True
    include_timestamp: bool = True
    footer_text: str
    header_text: str
    watermark_text: str | None = None


class ReportSettingsUpdate(BaseSchema):
    """Schema for updating report settings."""

    company_name: str | None = None
    company_logo: str | None = None
    primary_color: str | None = None
    secondary_color: str | None = None
    default_font: str | None = None
    default_page_size: str | None = Field(None, pattern="^(letter|a4|legal)$")
    default_orientation: str | None = Field(None, pattern="^(portrait|landscape)$")
    include_page_numbers: bool | None = None
    include_table_of_contents: bool | None = None
    include_timestamp: bool | None = None
    footer_text: str | None = None
    header_text: str | None = None
    watermark_text: str | None = None

File: src/lib/api/schemas/reporting.ts
/**
 * Zod schemas for the /reporting/* API surface
 *
 * Validates snake_case JSON responses and transforms to camelCase.
 * Replaces snakeToCamel() and provides runtime validation for templates,
 * queued reports, distribution schedules, widgets, and settings.
 */
import { z } from 'zod';

// ---------- Shared enum constants ----------

export const REPORT_CATEGORIES = ['executive', 'financial', 'market', 'portfolio', 'custom'] as const;
export const REPORT_FORMATS = ['pdf', 'excel', 'pptx'] as const;
export const REPORT_STATUSES = ['pending', 'generating', 'completed', 'failed'] as const;
export const SCHEDULE_FREQUENCIES = ['daily', 'weekly', 'monthly', 'quarterly'] as const;

/**
 * Schema for the raw snake_case API response.
 * Transforms to camelCase ReportSettings on parse.
 */
export const reportSettingsResponseSchema = z
  .object({
    company_name: z.string(),
    company_logo: z.string().nullable(),
    primary_color: z.string(),
    secondary_color: z.string(),
    default_font: z.string(),
    default_page_size: z.enum(['letter', 'a4', 'legal']),
    default_orientation: z.enum(['portrait', 'landscape']),
    include_page_numbers: z.boolean(),
    include_table_of_contents: z.boolean(),
    include_timestamp: z.boolean(),
    footer_text: z.string(),
    header_text: z.string(),
    watermark_text: z.string().nullable(),
  })
  .transform((s) => ({
    companyName: s.company_name,
    companyLogo: s.company_logo ?? undefined,
    primaryColor: s.primary_color,
    secondaryColor: s.secondary_color,
    defaultFont: s.default_font,
    defaultPageSize: s.default_page_size,
    defaultOrientation: s.default_orientation,
    includePageNumbers: s.include_page_numbers,
    includeTableOfContents: s.include_table_of_contents,
    includeTimestamp: s.include_timestamp,
    footerText: s.footer_text,
    headerText: s.header_text,
    watermarkText: s.watermark_text ?? undefined,
  }));

// ---------- Report Template ----------

export const reportTemplateResponseSchema = z
  .object({
    id: z.number(),
    name: z.string(),
    description: z.string().nullable().optional(),
    category: z.enum(REPORT_CATEGORIES),
    sections: z.array(z.string()),
    export_formats: z.array(z.enum(REPORT_FORMATS)),
    is_default: z.boolean(),
    created_by: z.string(),
    config: z.record(z.string(), z.unknown()).nullable().optional(),
    created_at: z.string(),
    updated_at: z.string(),
  })
  .transform((t) => ({
    id: t.id,
    name: t.name,
    description: t.description ?? undefined,
    category: t.category,
    sections: t.sections,
    exportFormats: t.export_formats,
    isDefault: t.is_default,
    createdBy: t.created_by,
    config: t.config ?? undefined,
    createdAt: t.created_at,
    updatedAt: t.updated_at,
  }));

export const reportTemplateListResponseSchema = z
  .object({
    items: z.array(reportTemplateResponseSchema),
    total: z.number(),
    page: z.number(),
    page_size: z.number(),
  })
  .transform((r) => ({
    items: r.items,
    total: r.total,
    page: r.page,
    pageSize: r.page_size,
  }));

// ---------- Queued Report ----------

export const queuedReportResponseSchema = z
  .object({
    id: z.number(),
    name: z.string(),
    template_id: z.number(),
    template_name: z.string().nullable().optional(),
    format: z.enum(REPORT_FORMATS),
    requested_by: z.string(),
    status: z.enum(REPORT_STATUSES),
    progress: z.number(),
    requested_at: z.string(),
    completed_at: z.string().nullable().optional(),
    file_size: z.string().nullable().optional(),
    download_url: z.string().nullable().optional(),
    error: z.string().nullable().optional(),
    created_at: z.string(),
    updated_at: z.string(),
  })
  .transform((q) => ({
    id: q.id,
    name: q.name,
    templateId: q.template_id,
    templateName: q.template_name ?? undefined,
    format: q.format,
    requestedBy: q.requested_by,
    status: q.status,
    progress: q.progress,
    requestedAt: q.requested_at,
    completedAt: q.completed_at ?? undefined,
    fileSize: q.file_size ?? undefined,
    downloadUrl: q.download_url ?? undefined,
    error: q.error ?? undefined,
    createdAt: q.created_at,
    updatedAt: q.updated_at,
  }));

export const queuedReportListResponseSchema = z
  .object({
    items: z.array(queuedReportResponseSchema),
    total: z.number(),
    page: z.number(),
    page_size: z.number(),
  })
  .transform((r) => ({
    items: r.items,
    total: r.total,
    page: r.page,
    pageSize: r.page_size,
  }));

// ---------- Distribution Schedule ----------

export const distributionScheduleResponseSchema = z
  .object({
    id: z.number(),
    name: z.string(),
    template_id: z.number(),
    template_name: z.string().nullable().optional(),
    recipients: z.array(z.string()),
    frequency: z.enum(SCHEDULE_FREQUENCIES),
    day_of_week: z.number().nullable().optional(),
    day_of_month: z.number().nullable().optional(),
    time: z.string(),
    format: z.enum(REPORT_FORMATS),
    is_active: z.boolean(),
    last_sent: z.string().nullable().optional(),
    next_scheduled: z.string(),
    created_at: z.string(),
    updated_at: z.string(),
  })
  .transform((d) => ({
    id: d.id,
    name: d.name,
    templateId: d.template_id,
    templateName: d.template_name ?? undefined,
    recipients: d.recipients,
    frequency: d.frequency,
    dayOfWeek: d.day_of_week ?? undefined,
    dayOfMonth: d.day_of_month ?? undefined,
    time: d.time,
    format: d.format,
    isActive: d.is_active,
    lastSent: d.last_sent ?? undefined,
    nextScheduled: d.next_scheduled,
    createdAt: d.created_at,
    updatedAt: d.updated_at,
  }));

export const distributionScheduleListResponseSchema = z.object({
  items: z.array(distributionScheduleResponseSchema),
  total: z.number(),
});

// ---------- Generate Report ----------

export const generateReportRequestSchema = z.object({
  template_id: z.number(),
  name: z.string(),
  format: z.enum(REPORT_FORMATS),
  parameters: z.record(z.string(), z.unknown()).nullable().optional(),
});

export const generateReportResponseSchema = z
  .object({
    queued_report_id: z.number(),
    status: z.enum(REPORT_STATUSES),
    message: z.string(),
  })
  .transform((g) => ({
    queuedReportId: g.queued_report_id,
    status: g.status,
    message: g.message,
  }));

// ---------- Report Widget ----------

export const reportWidgetResponseSchema = z
  .object({
    id: z.string(),
    type: z.string(),
    name: z.string(),
    description: z.string(),
    category: z.string(),
    icon: z.string(),
    default_width: z.number(),
    default_height: z.number(),
    configurable: z.boolean(),
  })
  .transform((w) => ({
    id: w.id,
    type: w.type,
    name: w.name,
    description: w.description,
    category: w.category,
    icon: w.icon,
    defaultWidth: w.default_width,
    defaultHeight: w.default_height,
    configurable: w.configurable,
  }));

export const reportWidgetListResponseSchema = z.object({
  widgets: z.array(reportWidgetResponseSchema),
  total: z.number(),
});
